scrapy/utils/conf.py

```diff
@@ -1,36 +1,49 @@
 import os
 import sys
+import warnings
 from operator import itemgetter
 
 import six
 from six.moves.configparser import SafeConfigParser
 
+from scrapy.settings import BaseSettings
 from scrapy.utils.deprecate import update_classpath
 
 
-def build_component_list(base, custom, convert=update_classpath):
-    """Compose a component list based on a custom and base dict of components
-    (typically middlewares or extensions), unless custom is already a list, in
-    which case it's returned.
-    """
+def build_component_list(compdict, convert=update_classpath):
+    """Compose a component list from a { class: order } dictionary."""
 
     def _check_components(complist):
         if len({convert(c) for c in complist}) != len(complist):
             raise ValueError('Some paths in {!r} convert to the same object, '
                              'please update your settings'.format(complist))
 
-    if isinstance(custom, (list, tuple)):
-        _check_components(custom)
-        return type(custom)(convert(c) for c in custom)
-
     def _map_keys(compdict):
+        if isinstance(compdict, BaseSettings):
+            compbs = BaseSettings()
+            for k, v in six.iteritems(compdict):
+                prio = compdict.getpriority(k)
+                if compbs.getpriority(convert(k)) == prio:
+                    raise ValueError('Some paths in {!r} convert to the same'
+                                     'object, please update your settings'
+                                     ''.format(compdict))
```

Comment:
Minor detail: Printing `{!r}` for BaseSettings is probably a little bit too verbose, since it prints the SettingsAttributes `repr`s. We could print `list(compdict.keys())` instead (The `list()` call is to convert the `dict_keys` objects in python3 to a proper list).